File: test/historical_test.dart
// Historical Intelligence — the Dart twin of site/lib/historical.test.ts.
// Confirms the reconstruction never fabricates a point it has no real
// anchor for, and that coverage/turning-points/comparisons degrade to
// explicit gaps rather than guesses.
import 'package:flutter_test/flutter_test.dart';

import 'package:crititrack/core/domain/models/celebrity.dart';
import 'package:crititrack/core/domain/models/controversy.dart';
import 'package:crititrack/core/domain/models/person_facts.dart';
import 'package:crititrack/core/domain/models/sentiment_data.dart';
import 'package:crititrack/core/utils/changes.dart';
import 'package:crititrack/core/utils/claims.dart';
import 'package:crititrack/core/utils/controversy_index.dart';
import 'package:crititrack/core/utils/coverage.dart';
import 'package:crititrack/core/utils/historical.dart';

CareerEntry careerEntry({
  String? role = 'Analyst',
  String? organization = 'Firm A',
  int? start = 2020,
  int? end,
}) => CareerEntry(
  role: role,
  organization: organization,
  start: start,
  end: end,
  sourceName: 'Wikidata',
  sourceUrl: 'https://www.wikidata.org/wiki/Q1',
);

Controversy controversy({
  String title = 'Fraud allegations',
  int severity = 4,
  int? year = 2021,
  List<String> sources = const ['https://reuters.com/1'],
}) => Controversy(
  title: title,
  summary: 'The executive was accused of misrepresenting finances.',
  category: ControversyCategory.financial,
  severity: severity,
  status: ControversyStatus.ongoing,
  year: year,
  sources: sources,
);

SentimentSnapshot snapshot(String date, double score, int mentions) => SentimentSnapshot(
  date: date,
  positiveCount: 0,
  negativeCount: 0,
  neutralCount: 0,
  totalMentions: mentions,
  dominantEmotion: 'neutral',
  score: score,
);

Celebrity celebrity({
  List<CareerEntry> career = const [],
  List<Controversy> controversies = const [],
  List<SentimentSnapshot> trend = const [],
}) => Celebrity(
  slug: 'jane-doe',
  name: 'Jane Doe',
  candidates: const [],
  biography: Biography(
    profession: 'Executive',
    summary: 'Jane Doe is an executive.',
    background: '',
    notableWorks: const [],
    controversies: controversies,
  ),
  sentimentData: SentimentData(
    overallScore: 50,
    positiveRatio: 0.3,
    negativeRatio: 0.3,
    neutralRatio: 0.4,
    trendDirection: 'stable',
    explanation: '',
    trendData: trend,
    dominantEmotion: 'neutral',
    sampleSize: 50,
    confidence: 0.8,
  ),
  mediaItems: const [],
  fetchedAt: DateTime.utc(2026, 9, 5),
  facts: PersonFacts(career: career),
);

void main() {
  group('buildHistoricalSnapshots', () {
    test('returns nothing with fewer than two measured points', () {
      final c = celebrity(trend: [snapshot('2024-01-01', 50, 10)]);
      expect(buildHistoricalSnapshots(c, const []), isEmpty);
    });

    test('builds one snapshot per measured sentiment date, sorted ascending', () {
      final c = celebrity(
        trend: [snapshot('2024-02-01', 60, 20), snapshot('2024-01-01', 50, 10)],
      );
      final snaps = buildHistoricalSnapshots(c, const []);
      expect(snaps.map((s) => s.capturedAt).toList(), ['2024-01-01', '2024-02-01']);
      expect(snaps.first.sentimentScore, 50);
    });

    test('overlays career state as of each snapshot year — no role before its start date', () {
      final c = celebrity(
        trend: [snapshot('2019-06-01', 50, 5), snapshot('2021-06-01', 55, 8)],
        career: [careerEntry(start: 2020, role: 'Analyst', organization: 'Firm A')],
      );
      final snaps = buildHistoricalSnapshots(c, const []);
      expect(snaps[0].currentRole, isNull);
      expect(snaps[1].currentRole, 'Analyst, Firm A');
    });

    test('counts only controversies dated on or before the snapshot year', () {
      final c = celebrity(
        trend: [snapshot('2020-06-01', 50, 5), snapshot('2022-06-01', 40, 5)],
        controversies: [controversy(year: 2021)],
      );
      final snaps = buildHistoricalSnapshots(c, const []);
      expect(snaps[0].controversyCount, 0);
      expect(snaps[1].controversyCount, 1);
      expect(snaps[0].critiScore, isNull);
      expect(snaps[1].critiScore, isNotNull);
    });
  });

  group('supportedTimeRanges / filterSnapshotsByRange', () {
    test('offers no ranges with fewer than two snapshots', () {
      expect(supportedTimeRanges(const []), isEmpty);
    });

    test('only offers ranges the data span actually supports, plus all', () {
      final now = DateTime.utc(2024, 3, 1);
      final c = celebrity(
        trend: [snapshot('2024-02-05', 50, 5), snapshot('2024-02-25', 55, 5)],
      );
      final snaps = buildHistoricalSnapshots(c, const []);
      final ranges = supportedTimeRanges(snaps, now: now);
      expect(ranges, contains(HistoricalTimeRange.d30));
      expect(ranges, contains(HistoricalTimeRange.all));
      expect(ranges, isNot(contains(HistoricalTimeRange.y5)));
    });
  });

  group('buildHistoricalCoverage', () {
    test('marks every dimension unavailable, never zero, with no data', () {
      final cov = buildHistoricalCoverage(const [], const [], const [], const [], const []);
      for (final d in cov) {
        expect(d.level, CoverageLevel.unavailable);
        expect(d.status, DataStatus.unavailable);
      }
    });
  });

  group('majorTurningPoints', () {
    test('flags only large year-over-year CritiScore reconstruction moves', () {
      final points = majorTurningPoints(
        const [
          IndexHistoryPoint(year: 2020, score: 10),
          IndexHistoryPoint(year: 2021, score: 12),
          IndexHistoryPoint(year: 2022, score: 60),
        ],
        const [],
        const [],
      );
      expect(points.any((p) => p.kind == TurningPointKind.score && p.date == '2022'), isTrue);
      expect(points.any((p) => p.date == '2021'), isFalse);
    });
  });

  group('comparePeriods', () {
    test('a period with no snapshots reports nulls, not fabricated zeros', () {
      final cmp = comparePeriods(const [], HistoricalTimeRange.d30, HistoricalTimeRange.y1);
      expect(cmp.startScoreA, isNull);
      expect(cmp.endScoreA, isNull);
      expect(cmp.controversyCountA, 0);
    });
  });

  group('buildHistoricalOverview', () {
    test('hasHistory is false with no snapshots, no score history, no turning points', () {
      final overview = buildHistoricalOverview(
        celebrity: celebrity(),
        claims: const <Claim>[],
        changeEvents: const <ChangeEvent>[],
      );
      expect(overview.hasHistory, isFalse);
      expect(overview.firstSnapshotDate, isNull);
    });

    test('hasHistory is true once real dated controversies produce a score history', () {
      final c = celebrity(
        controversies: [controversy(year: 2019), controversy(title: 'Second', year: 2022)],
      );
      final overview = buildHistoricalOverview(
        celebrity: c,
        claims: const <Claim>[],
        changeEvents: const <ChangeEvent>[],
      );
      expect(overview.hasHistory, isTrue);
    });
  });
}

File: site/lib/historical.test.ts
import { describe, expect, test } from "vitest";
import {
  buildHistoricalSnapshots,
  buildHistoricalCoverage,
  buildHistoricalOverview,
  supportedTimeRanges,
  filterSnapshotsByRange,
  majorTurningPoints,
  comparePeriods,
  filterTurningPoints,
} from "./historical";
import type { Controversy } from "./controversy";
import type { CareerEntry } from "./career";
import type { Claim } from "./claims";
import type { ChangeEvent } from "./changes";
import type { TrendPoint } from "./api";

function careerEntry(overrides: Partial<CareerEntry> = {}): CareerEntry {
  return {
    start: 2020,
    end: null,
    role: "Analyst",
    organization: "Firm A",
    location: null,
    industry: null,
    current: true,
    source: { name: "Wikidata", url: "https://www.wikidata.org/wiki/Q1" },
    ...overrides,
  };
}

function controversy(overrides: Partial<Controversy> = {}): Controversy {
  return {
    title: "Fraud allegations",
    summary: "The executive was accused of misrepresenting finances.",
    category: "Financial",
    severity: 4,
    status: "ongoing",
    year: 2021,
    sources: ["https://reuters.com/1"],
    ...overrides,
  };
}

function trend(points: Array<[string, number, number]>): TrendPoint[] {
  return points.map(([date, score, mentions]) => ({ date, score, mentions }));
}

function baseProfile(overrides: {
  trend?: TrendPoint[];
  controversies?: Controversy[];
  career?: CareerEntry[];
} = {}) {
  return {
    slug: "jane-doe",
    trend: overrides.trend ?? [],
    controversies: overrides.controversies ?? [],
    career: { timeline: overrides.career ?? [], organizations: [], insights: { start: null, current: null, transitions: [], leadershipRoles: [], founder: false, progression: [] }, available: (overrides.career ?? []).length > 0 },
  };
}

describe("buildHistoricalSnapshots", () => {
  test("returns nothing with fewer than two measured points — a single point has no shape", () => {
    const profile = baseProfile({ trend: trend([["2024-01-01", 50, 10]]) });
    expect(buildHistoricalSnapshots(profile, [])).toEqual([]);
  });

  test("builds one snapshot per measured sentiment date, sorted ascending", () => {
    const profile = baseProfile({
      trend: trend([
        ["2024-02-01", 60, 20],
        ["2024-01-01", 50, 10],
      ]),
    });
    const snaps = buildHistoricalSnapshots(profile, []);
    expect(snaps.map((s) => s.capturedAt)).toEqual(["2024-01-01", "2024-02-01"]);
    expect(snaps[0].sentimentScore).toBe(50);
  });

  test("overlays career state as of each snapshot's year — no role before its start date", () => {
    const profile = baseProfile({
      trend: trend([
        ["2019-06-01", 50, 5],
        ["2021-06-01", 55, 8],
      ]),
      career: [careerEntry({ start: 2020, role: "Analyst", organization: "Firm A" })],
    });
    const snaps = buildHistoricalSnapshots(profile, []);
    expect(snaps[0].currentRole).toBeNull();
    expect(snaps[1].currentRole).toBe("Analyst, Firm A");
  });

  test("counts only controversies dated on or before the snapshot's year", () => {
    const profile = baseProfile({
      trend: trend([
        ["2020-06-01", 50, 5],
        ["2022-06-01", 40, 5],
      ]),
      controversies: [controversy({ year: 2021 })],
    });
    const snaps = buildHistoricalSnapshots(profile, []);
    expect(snaps[0].controversyCount).toBe(0);
    expect(snaps[1].controversyCount).toBe(1);
    expect(snaps[0].critiScore).toBeNull();
    expect(snaps[1].critiScore).not.toBeNull();
  });

  test("counts claims that belong to a controversy in scope, keyed by titleSlug", () => {
    const c = controversy({ title: "Fraud allegations", year: 2021 });
    const claim: Claim = {
      claimId: "c1",
      entityId: null,
      controversyId: "fraud-allegations",
      timelineEventId: null,
      claimText: "It happened",
      claimType: "allegation",
      dateContext: "2021",
      status: "unverified",
      confidence: "low",
      supportingEvidenceIds: [],
      contradictingEvidenceIds: [],
      neutralEvidenceIds: [],
      responseEvidenceIds: [],
      createdAt: null,
      updatedAt: null,
      methodologyVersion: "cvm-1",
    } as unknown as Claim;
    const profile = baseProfile({ trend: trend([["2021-01-01", 40, 1], ["2022-01-01", 40, 1]]), controversies: [c] });
    const snaps = buildHistoricalSnapshots(profile, [claim]);
    expect(snaps[1].claimCount).toBe(1);
  });
});

describe("supportedTimeRanges / filterSnapshotsByRange", () => {
  test("offers no ranges with fewer than two snapshots", () => {
    expect(supportedTimeRanges([])).toEqual([]);
  });

  test("only offers ranges the data span actually supports, plus 'all'", () => {
    const now = Date.parse("2024-03-01T00:00:00Z");
    const profile = baseProfile({
      trend: trend([
        ["2024-02-05", 50, 5],
        ["2024-02-25", 55, 5],
      ]),
    });
    const snaps = buildHistoricalSnapshots(profile, []);
    const ranges = supportedTimeRanges(snaps, now);
    expect(ranges).toContain("30d");
    expect(ranges).toContain("all");
    expect(ranges).not.toContain("5y");
  });

  test("filterSnapshotsByRange('all') returns everything unfiltered", () => {
    const profile = baseProfile({ trend: trend([["2010-01-01", 50, 1], ["2024-01-01", 60, 1]]) });
    const snaps = buildHistoricalSnapshots(profile, []);
    expect(filterSnapshotsByRange(snaps, "all").length).toBe(2);
  });
});

describe("buildHistoricalCoverage", () => {
  test("marks every dimension unavailable, never zero, when there is no data", () => {
    const cov = buildHistoricalCoverage([], [], [], [], []);
    for (const d of cov) {
      expect(d.level).toBe("unavailable");
      expect(d.status).toBe("unavailable");
    }
  });
});

describe("majorTurningPoints", () => {
  test("flags only large year-over-year CritiScore reconstruction moves", () => {
    const points = majorTurningPoints(
      [
        { year: 2020, score: 10 },
        { year: 2021, score: 12 },
        { year: 2022, score: 60 },
      ],
      [],
      [],
    );
    expect(points.some((p) => p.kind === "score" && p.date === "2022")).toBe(true);
    expect(points.some((p) => p.date === "2021")).toBe(false);
  });

  test("includes MAJOR/SIGNIFICANT change events but not MINOR/INFO ones", () => {
    const change = (severity: ChangeEvent["severity"], id: string): ChangeEvent => ({
      changeId: id,
      entityId: "jane-doe",
      changeType: "CAREER_CHANGE",
      severity,
      title: `Change ${id}`,
      summary: "",
      previousValue: null,
      currentValue: null,
      detectedAt: "2024-01-01T00:00:00Z",
      effectiveDate: null,
      evidenceIds: [],
      relatedClaimIds: [],
      methodologyVersion: "1.0",
      confidence: "HIGH",
      sourceCoverage: null,
    });
    const points = majorTurningPoints([], [], [change("MAJOR", "a"), change("MINOR", "b")]);
    expect(points.some((p) => p.id === "a")).toBe(true);
    expect(points.some((p) => p.id === "b")).toBe(false);
  });
});

describe("filterTurningPoints", () => {
  test("'all' returns everything; a kind filters to only that kind", () => {
    const points = [
      { id: "1", kind: "score" as const, date: "2020", title: "t", summary: "s" },
      { id: "2", kind: "career" as const, date: "2021", title: "t", summary: "s" },
    ];
    expect(filterTurningPoints(points, "all").length).toBe(2);
    expect(filterTurningPoints(points, "career")).toEqual([points[1]]);
  });
});

describe("comparePeriods", () => {
  test("a period with no snapshots reports nulls, not fabricated zeros", () => {
    const cmp = comparePeriods([], "30d", "1y");
    expect(cmp.startScoreA).toBeNull();
    expect(cmp.endScoreA).toBeNull();
    expect(cmp.controversyCountA).toBe(0);
  });
});

describe("buildHistoricalOverview", () => {
  test("hasHistory is false for a figure with no snapshots, no score history, no turning points", () => {
    const overview = buildHistoricalOverview({ profile: baseProfile(), claims: [], changeEvents: [] });
    expect(overview.hasHistory).toBe(false);
    expect(overview.firstSnapshotDate).toBeNull();
  });

  test("hasHistory is true once real dated controversies produce a score history", () => {
    const profile = baseProfile({
      controversies: [controversy({ year: 2019 }), controversy({ title: "Second", year: 2022 })],
    });
    const overview = buildHistoricalOverview({ profile, claims: [], changeEvents: [] });
    expect(overview.hasHistory).toBe(true);
  });
});
